spyder/plugins/ipythonconsole/widgets/kernel_connection_manager.py

```diff
@@ -0,0 +1,409 @@
+        try:
+            self.kernel_manager.shutdown_kernel()
+        except Exception:
+            # kernel was externally killed
+            pass
+
+    def wait_shutdown_thread(self):
+        """Wait shutdown thread."""
+        thread = self.shutdown_thread
+        if thread is None:
+            return
+        if thread.isRunning():
+            try:
+                thread.kernel_manager._kill_kernel()
+            except Exception:
+                pass
+            thread.quit()
+            thread.wait()
+
+    def copy(self):
+        """Copy kernel."""
+        # Copy kernel infos
+        new_kernel = self.__class__(
+            connection_file=self.connection_file,
+            kernel_manager=self.kernel_manager,
+            known_spyder_kernel=self.known_spyder_kernel,
+            hostname=self.hostname,
+            sshkey=self.sshkey,
+            password=self.password,
+        )
+
+        # Copy std file
+        if self.stderr_obj is not None:
+            new_kernel.stderr_obj = self.stderr_obj.copy()
+        if self.stdout_obj is not None:
+            new_kernel.stdout_obj = self.stdout_obj.copy()
+        if self.fault_obj is not None:
+            new_kernel.fault_obj = self.fault_obj.copy()
+
+        # Get new kernel_client
+        new_kernel.init_kernel_client()
+        return new_kernel
+
+    def remove_files(self):
+        """Remove std files."""
+        for obj in [self.stderr_obj, self.stderr_obj, self.fault_obj]:
+            if obj is not None:
+                obj.remove()
+
+    def open_comm(self, kernel_comm):
+        """Open kernel comm"""
+        kernel_comm.open_comm(self.kernel_client)
+        self.kernel_comm = kernel_comm
+
+    def replace_std_files(self):
+        """Replace std files."""
+        for obj in [self.stderr_obj, self.stderr_obj, self.fault_obj]:
+            if obj is None:
+                continue
+            obj.remove()
+            fn = obj.filename
+            m = re.match(r"(.+_)(\d+)(.[a-z]+)", fn)
+            if m:
+                # Already a replaced file
+                path, n, ext = m.groups()
+                obj.filename = path + str(1 + int(n)) + ext
+                continue
+            m = re.match(r"(.+)(.[a-z]+)", fn)
+            if m:
+                # First replaced file
+                path, ext = m.groups()
+                obj.filename = path + "_1" + ext
+                continue
+            # No extension, should not happen
+            obj.filename += "_1"
+
+
+class CachedKernelMixin:
+    """Cached kernel mixin."""
+
+    def __init__(self):
+        super().__init__()
+        self._cached_kernel_properties = None
+
+    def close_cached_kernel(self):
+        """Close the cached kernel."""
+        if self._cached_kernel_properties is None:
+            return
+        kernel = self._cached_kernel_properties[-1]
+        kernel.close(now=True)
+        kernel.remove_files()
```

Comment:
I think this `kernel` variable should be renamed to `kernel_handler`, or another name you consider more appropriate. That's because `kernel` could mean that it gives direct access to the kernel but that's not the case, which is a bit confusing.